### PR TITLE
SPV deposit sweep maintainer logic

### PR DESCRIPTION
**Pull request by tomaszslabon**
Refs: https://github.com/keep-network/keep-core/issues/3615.
This PR adds implementation of deposit sweep proofs in SPV maintainer.
The SPV maintainer periodically checks if there are unproven deposit sweep transactions.
If there are such transactions, it assembles the proof and submits it.

Example command:
```
KEEP_ETHEREUM_PASSWORD="password" ./keep-client maintainer --spv --spv.historyDepth 200000 --config maintainer.toml 
```

### CODE REVIEW

File: pkg/maintainer/spv/chain.go
Comment (pdyraga) [inline pkg/maintainer/spv/chain.go:33]: Agreed, those functions are mostly identical right now.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:64]: I would drop this line given we log earlier:

https://github.com/keep-network/keep-core/blob/8030414dee6ea771e69b21278163e8aee1cb6319/pkg/maintainer/spv/spv.go#L40

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:167]: Hmm... why? I can't think of a scenario when we should wait with providing SPV proof but of course, I could miss something. In my opinion, it should happen as soon as possible - if the wallet was late with signing, providing SPV proof right away will allow the next action to be performed.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:171]: In theory, the wallet should not execute the next action if the previous one was not yet completed. I think checking 5 as a default is fine, but I would also allow this value to be passed as a parameter.

Comment (pdyraga) [reply pkg/maintainer/spv/spv.go:171]: Oh, wait. There is a chance the wallet performed e.g. 20 redemptions after sweeping. This function fetches all Bitcoin transactions no matter the type.

Assuming we execute sweep every 48 hours and redemption every 3 hours, it seems that the default value should be closer to 20.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:248]:
```suggestion
				return false, fmt.Errorf("failed to check if input is the main UTXO")
```

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go:167]: We could filter out wallets that are not `Live` or `MovingFunds` for performance reasons.
If the wallet is `Closed`, we could skip it immediately.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:304]: IMO, we should return an error. Most likely, this is something bad on the infrastructure level (e.g. connection error to Ethereum node). Less likely, this is some terrible bug in the code.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [reply pkg/maintainer/spv/spv.go:167]: That's a good point, I thought this is about the wallet lock state in the `WalletCoordinator` contract. If this is easy, I think we should do it. This is not an issue now when all wallets are Live but we'll get to the point when this check will matter.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go:95]: And if it failed on-chain, the maintainer would keep trying to submit it anyway on each round of processing.
It would lead to unnecessary burning of funds from the `ReimbursementPool`.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: Done in 32150aba63c2cea884d25fa231532958169bdcb1.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: Done in 28e3faa5fa4eda8ca3f2f451f60ac5889aa2cd40.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: Agree. Left the code as is (we return an error if wallet not found) and removed the TODO in ae31a234b0f5ba60571f63479c158d633c15a4eb.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go:192]: Added the `HistoryDepth` parameter in d2e4466b6f5a17f3c8854c280472c2110771cf1e.
The default value for it is `40320`, the approximate number of Ethereum blocks in one week:
7 * 24 * 3600 / 15 = 40320

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: Done in 45bb022c94708fe000003e3ea54112c81f967cab.
The default value for the parameter is `20`.

File: pkg/maintainer/spv/chain.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/chain.go:34]: As joining `GetDepositRequest` and `Deposits` modifies some code unrelated to this PR I decided to handle it in a small separate PR: https://github.com/keep-network/keep-core/pull/3654.
Once the PR is merged to main, I will remove the `Deposits` function.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: Added a check for wallet state in 8abdeecdac4f3cc85fce44b60968bab6d68b7d94.

File: pkg/chain/ethereum/tbtc.go
Comment (pdyraga) [inline pkg/chain/ethereum/tbtc.go:1189]: Note to self: this should be removed after #3654.

File: pkg/maintainer/spv/config.go
Comment (pdyraga) [inline pkg/maintainer/spv/config.go:21]: Let's add some extra context for future-us and other contributors. The values of those two parameters are crucial. My proposition:

```suggestion
	// HistoryDepth is the number of blocks to look back from the current block
	// when searching for past deposit sweep proposal submitted events.
	// To find Bitcoin transactions for which the SPV proof should be submitted,
	// the maintainer first inspects the sweep proposal events from the wallet
	// coordinator contract. This depth determines how far into the past the
	// system will consider events for processing. This value must not be too
	// high so that the event lookup is efficient. At the same time, this value
	// can not be too low to make sure all performed and not yet proven sweeps
	// can be found.
	HistoryDepth uint64

	// TransactionLimit sets the maximum number of confirmed transactions
	// returned when getting transactions for a public key hash. Once the
	// maintainer establishes the list of proposed sweeps based on the sweep
	// proposal events from the wallet coordinator contract, it needs to check
	// Bitcoin transactions executed by the wallet. Then, it tries to find the
	// sweep among those transactions. For example, if set to `20`, only the
	// latest twenty transactions will be returned. This value must not be too
	// high so that the transaction lookup is efficient. At the same time, this
	// value can not be too low to make sure the performed sweep can be found in
	// case the wallet decided to execute some other Bitcoin transaction after
	// the yet-not-proven sweep.
	TransactionLimit int
```

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:29]: We could also mention this small yet important wallet implementation detail.

One could argue we shouldn't mention wallet implementation details in the maintainer docs and there is a truth in it but... this is quite a crucial piece of knowledge and I think it's worth mentioning here in case we need to make some changes in the SPV maintainer logic in the future.

```suggestion
	// Default value for the limit of transactions returned for a given wallet
	// public key hash. The value is based on the frequency of how often deposit
	// sweep and redemption transaction will happen. Deposit sweep transactions
	// are assumed to happen every 48h. Redemption transactions are assumed to
	// happen every 3h. The wallet should refuse any proposals from the
	// coordinator if the previously executed Bitcoin transaction was not proved
	// to the Bridge yet so in theory, the value of 1 should be enough. We make
	// it a bit higher - better to be safe than sorry.
```

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:100]: I really like how the control flow is organized in those functions. The `for` loops, the `select`s, and how the functions are divided. It makes a really logical structure that is easy to follow. 👌🏻

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:233]: The values of `1` and `2` are Solidity/EVM implementation details. I think we should modify `WalletChainData` structure to return `State` not as `uint8` but as `WalletState` enum we will introduce. We have `WalletActionType` and `parseWalletActionType` in the Ethereum implementation. We should introduce identical enum for `WalletState` and have an identical parsing function. I don't think we should necessarily reuse the same code given `WalletActionType` comes from the `WalletCoordinator` contract and it is another type.

File: pkg/chain/ethereum/tbtc.go
Comment (tomaszslabon) [reply pkg/chain/ethereum/tbtc.go]: Removed in cbd754abb9a1718e495209884eb4714db6e5ad68.

File: pkg/maintainer/spv/config.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/config.go]: Done in 72058add43edb65b5c2176770366dcc99ced7685.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: Done in 72058add43edb65b5c2176770366dcc99ced7685.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go:256]: Done in 2386f5ba3d386d0db1144f88dc4b76631dc9901b.
Unfortunately, I had to prefix each state constant:
```
const (
	StateUnknown WalletState = iota
	StateLive
	StateMovingFunds
	StateClosing
	StateClosed
	StateTerminated
)
``` 
with `State...`, because we already have `MovingFunds` constant in `WalletActionType`.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: I'll do the TODO in a follow-up PR.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [reply pkg/maintainer/spv/spv.go]: I propose that we prefix all types of `WalletActionType` with `Action*` in a separate PR to keep it consistent.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:24]: Shouldn't it be 50400 assuming 12s block time? 🤔

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:182]: What do you think about moving this piece of code to a function exposed by `bitcoin.SpvProof`? We'd use it here as `proof.FirstBlockHeaderDifficulty()`, for example.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:205]: I think we should log a warning in this case. The second case is pretty dangerous. 

We should also make sure those warnings are included in our monitoring rules cc @lukasz-zimnoch.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:285]: Let's add an info-level log for this case.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:162]: Let's add an info-level log for this case.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: Done in a separate PR https://github.com/keep-network/keep-core/pull/3662.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: It should. For some reason I assumed one Ethereum block is about `15s`, but it is indeed close to `12s` (at least in recent months).
Done in 004af961764e3fa51e63aecfa0485b9a32e1990e.

File: pkg/chain/ethereum/bitcoin_difficulty.go
Comment (tomaszslabon) [reply pkg/chain/ethereum/bitcoin_difficulty.go]: Done in e08ef679f7f3ac2f32982ed52917c4dc1fd5e024.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: It makes sense. Done in 7cd1de0b8dbe6cc5ed8db4c8854ad2f46484d89e.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go:214]: Done in 132ac7d666c8964b47a71255b3c50334f240fe22.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go:305]: Done in 132ac7d666c8964b47a71255b3c50334f240fe22.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go:174]: Done in 132ac7d666c8964b47a71255b3c50334f240fe22.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:74]: Let's align it with how we log the activity for the wallet coordinator maintainer:

```suggestion
		logger.Infof("starting deposit sweep proof task execution...")

		if err := sm.proveDepositSweepTransactions(); err != nil {
			return fmt.Errorf(
				"error while proving deposit sweep transactions: [%v]",
				err,
			)
		}

		logger.Infof(
			"deposit sweep proof task run completed; next run in [%s]",
			sm.config.IdleBackOffTime,
		)
```

Compare:

https://github.com/keep-network/keep-core/blob/fc1bbf3974f4562f65d74208f11c976fc04fe0d7/pkg/maintainer/wallet/wallet.go#L69-L96

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:90]: We should start log messages from a lower-case:

```suggestion
		"found [%d] unproven deposit sweep transaction(s)",
```

Let's align all other log messages in this file.

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:148]: Let's be more specific:

```suggestion
			// Skip the transaction as it has not accumulated enough
			// confirmations. It will be proven later.
			logger.Infof(
				"Skipped proving deposit sweep transaction [%s]; transaction "+
					"has [%v/%v] confirmations",
				transactionHashStr,
				accumulatedConfirmations,
				requiredConfirmations,
			)
```

File: pkg/maintainer/spv/spv.go
Comment (pdyraga) [inline pkg/maintainer/spv/spv.go:179]: Let's be more specific:

```suggestion
			logger.Warnf(
				"skipped proving deposit sweep transaction [%s]; found "+
					"difficulties mismatch between proof and relay: "+
					"proof first header difficulty: [%v], "+
					"previous relay difficulty: [%v], "+
					"current relay difficultry: [%v]",
				transactionHashStr,
				firstBlockHeaderDifficulty,
				previousDifficulty,
				currentDifficulty,
			)
```

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go:76]: Done in 61e77e4760e35d764dd85340f0251116f7ccc49f.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: Done in 69922dff13522866329c1ca88f804e025ecefb2c.

**Review by pdyraga — APPROVED** [review]

**Comment by pdyraga** [comment]
We agreed to implement unit tets for this code in a separate PR so that we can work on the remaining SPV maintainer logic in parallel. 

The remaining work is captured on the tasks list in https://github.com/keep-network/keep-core/issues/3615.

I tested this code locally and it submitted deposit sweep SPV proof successfully.

File: pkg/maintainer/spv/spv.go
Comment (tomaszslabon) [reply pkg/maintainer/spv/spv.go]: The followup PR: https://github.com/keep-network/keep-core/pull/3667.

**Pull request merged**
